ui/js/controllers/tcjobactions.js

```diff
@@ -0,0 +1,91 @@
+"use strict";
+
+treeherder.controller('TCJobActionsCtrl', [
+    '$scope', '$http', '$uibModalInstance', 'ThResultSetStore',
+    'ThJobDetailModel', 'thTaskcluster', 'ThTaskclusterErrors',
+    'thNotify', 'job', 'repoName', 'resultsetId',
+    function($scope, $http, $uibModalInstance, ThResultSetStore,
+             ThJobDetailModel, thTaskcluster, ThTaskclusterErrors, thNotify,
+             job, repoName, resultsetId) {
+        let originalTask;
+        $scope.input = {};
+
+        $scope.cancel = function() {
+            $uibModalInstance.dismiss('cancel');
+        };
+
+        $scope.updateSelectedAction = function() {
+            $scope.input.jsonPayload = JSON.stringify(jsonSchemaDefaults($scope.input.selectedAction.schema), null, 4);
+        };
+
+        $scope.triggerAction = function() {
+            $scope.triggering = true;
+
+            let tc = thTaskcluster.client();
+            let task = _.clone($scope.input.selectedAction.task);
+
+            ['created', 'deadline', 'expires'].map(function(dateProp) {
+                task[dateProp] = tc.fromNow(task[dateProp]['$fromNow']);
+            });
+            task.payload.env['ACTION_INPUT'] = $scope.input.jsonPayload;
+            task.payload.env['ACTION_PARAMETERS'] = null; // fixme: unclear what to set this to
+            task.payload.env['ACTION_TASK'] = null; // fixme: what is this supposed to do?
+            task.payload.env['ACTION_TASK_ID'] = JSON.stringify(job.taskcluster_metadata.task_id);
+            task.payload.env['ACTION_TASK_GROUP_ID'] = JSON.stringify(originalTask.taskGroupId);
```

Comment:
Assuming that `$scope.input.selectedAction` is an entry from the `actions` array of `actions.json`, then this doesn't even remotely implement the specification:
http://gecko.readthedocs.io/en/latest/taskcluster/taskcluster/action-spec.html#action-kind-task

```js
import _ from 'lodash';
import assert from 'assert';
import Ajv from 'ajv';

/////////////// Collect input

// Get static variables, this is the top-level 'variables' property from actions.json
let variables = $scope.actions_json.variables;

// Get the action wish to trigger, this is an entry from the top-level 'actions' array actions.json
let action = $scope.input.selectAction;

// Get input for the action
let input = JSON.parse($scope.input.jsonPayload || 'null'); // null if no input

// taskId of selected task we're triggering this action for, null if no task is selected 
let selectedTaskId = job.taskcluster_metadata.task_id || null;

// task definition of selected task, null if no task is selected
let selectedTask = null; 
if (selectedTaskId) {
  selectedTaskId = await (new tc.Queue()).task(selectedTaskId);
}

// Get the taskGroupId from any task in the result-set we're working on
let taskGroupId;
if (selectedTask) {
  taskGroupId = selectedTask.taskGroupId;
} else {
  taskGroupId = // TODO: Get the taskGroupId from the decision task, from the current result-set
}


/////////////// Sanity checks (UI shouldn't allow these to fail)

// this should never happen you should filter actions to ensure you only
// present supported actions to the user: actions = actions.filter(({kind) => kind === 'task');
assert(action.kind === 'task', 'Unsupported action kind');

// UI should now allows us to trigger an action _for a selected task_ if the
// action doesn't have a context. Ie. it's a global action, not a task specific action. 
assert((action.context && action.context.length > 0) || !selectedTaskId,
       'action has no context, thus accepts no selected task');

// Either we have a selectedTask and selectedTaskId, or neither
assert((selectedTaskId && selectedTask) || (!selectedTaskId && !selectedTask));

// UI should not allow us to trigger an action that isn't relevant for the selected task
assert(_.some(action.context || [], tagset => {
  // if there is some tagset in action.context, such that every key-value pair
  // from the tagset exists in task.tags, then the action is relevant
  return _.every(tagset, (val, key) => selectedTask.tags[key] === val);
}), 'action is not relevant for selectedTask');

// Validate input against schema, if input/schema is given
if (action.schema) {
  assert((new Ajv()).validate(action.schema, input), 'Input does not satisfy schema!');
} else {
  assert(input === null, 'no input is taken by this action, expected null!');
}


/////////////// Instantiate the task template

let render = // import from: https://gist.github.com/jonasfj/f4998c938bf85c5573d1359616bc086c

let actiontaskId = tc.slugid();
let actionTask = render(action.task, _.defaults({}, {
  taskGroupId: taskGroupId,
  taskId: selectedTaskId,
  task: selectedTask,
  input: input,
}, variables); // Important remember to supply static variables


/////////////// Create action task
let queue = tc.Queue({/* credentials from localStorage */});
await queue.createTask(actiontaskId, actionTask);
```